dev/PowerNotifications/PowerNotifications.cpp

```diff
@@ -0,0 +1,501 @@
+    DischargeTimeRegistration          PowerManager::m_dischargeTimeHandle;
+    PowerConditionRegistration         PowerManager::m_powerConditionHandle;
+    DisplayStatusRegistration          PowerManager::m_displayStatusHandle;
+    SystemIdleStatusRegistration       PowerManager::m_systemIdleStatusHandle;
+    PowerSchemePersonalityRegistration PowerManager::m_powerSchemePersonalityHandle;
+    UserPresenceStatusRegistration     PowerManager::m_userPresenceStatusHandle;
+    SystemAwayModeStatusRegistration   PowerManager::m_systemAwayModeStatusHandle;
+
+    std::mutex PowerManager::m_mutex;
+
+    int   PowerManager::m_batteryChargePercent, PowerManager::m_oldBatteryChargePercent;
+    DWORD PowerManager::m_cachedDisplayStatus;
+    DWORD PowerManager::m_cachedUserPresenceStatus;
+    DWORD PowerManager::m_cachedSystemAwayModeStatus;
+    DWORD PowerManager::m_cachedPowerCondition;
+    GUID  PowerManager::m_cachedPowerSchemePersonality;
+    ULONGLONG               PowerManager::m_cachedDischargeTime;
+    CompositeBatteryStatus *PowerManager::m_cachedCompositeBatteryStatus;
+    ::EnergySaverStatus PowerManager::m_cachedEnergySaverStatus;
+    winrt::Microsoft::ProjectReunion::BatteryStatus     PowerManager::m_batteryStatus;
+    winrt::Microsoft::ProjectReunion::BatteryStatus     PowerManager::m_oldBatteryStatus = winrt::Microsoft::ProjectReunion::BatteryStatus::Uninitialized;
+    winrt::Microsoft::ProjectReunion::PowerSupplyStatus PowerManager::m_powerSupplyStatus;
+    winrt::Microsoft::ProjectReunion::PowerSupplyStatus PowerManager::m_oldPowerSupplyStatus = winrt::Microsoft::ProjectReunion::PowerSupplyStatus::Uninitialized;
+
+
+    bool PowerManager::NotAlreadyRegisteredForEvents(PowerManager::eventType eventObj)
+    {
+        return !eventObj.get();
+    }
+
+    PowerManager::eventType PowerManager::GetEventObj(PowerFunction const& fn)
+    {
+        switch (fn) {
+        case EnergySaverStatusFn:      return PowerManager::m_energySaverStatusChangedEvent;
+        case BatteryStatusFn:          return PowerManager::m_batteryStatusChangedEvent;
+        case PowerSupplyStatusFn:      return PowerManager::m_powerSupplyStatusChangedEvent;
+        case RemainingChargePercentFn: return PowerManager::m_remainingChargePercentChangedEvent;
+        case RemainingDischargeTimeFn: return PowerManager::m_remainingDischargeTimeChangedEvent;
+        case PowerSourceStatusFn:      return PowerManager::m_powerSourceStatusChangedEvent;
+        case DisplayStatusFn:          return PowerManager::m_displayStatusChangedEvent;
+        case SystemIdleStatusFn:       return PowerManager::m_systemIdleStatusChangedEvent;
+        case PowerSchemePersonalityFn: return PowerManager::m_powerSchemePersonalityChangedEvent;
+        case UserPresenceStatusFn:     return PowerManager::m_userPresenceStatusChangedEvent;
+        case SystemAwayModeStatusFn:   return PowerManager::m_systemAwayModeStatusChangedEvent;
+        default: throw E_NOTIMPL;
```

Comment:
Don't throw raw HRESULTs. Use `throw winrt::hresult_not_implemented();` instead. Applies everywhere.

Reply:
(This is also developer/implementer error, so feel free to use WIL's FAIL_FAST_* family of macros.)